### PR TITLE
Remove ContainsKey where relevant

### PR DESCRIPTION
**Pull request by Forgind**
Each instance of ContainsKey that I deleted was followed by querying the same dictionary for the same key. These can be done at the same time, so I replaced them.

I do not expect that this will have any substantial user impact, but it theoretically improves perf, and it's a pet peeve of mine.

### CODE REVIEW

File: src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs [hunk 1]
```diff
@@ -484,10 +484,7 @@ internal bool AcquireAndSetUpHost(HandshakeOptions hostContext, INodePacketFacto
         /// </summary>
         internal void DisconnectFromHost(HandshakeOptions hostContext)
         {
-            ErrorUtilities.VerifyThrow(_nodeIdToPacketFactory.ContainsKey((int)hostContext) && _nodeIdToPacketHandler.ContainsKey((int)hostContext), "Why are we trying to disconnect from a context that we already disconnected from?  Did we call DisconnectFromHost twice?");
-
-            _nodeIdToPacketFactory.Remove((int)hostContext);
```
Comment (Forgind) [inline src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs:489]: This change technically means removing hostContext isn't atomic, but I don't think that's an issue.

File: src/Tasks/Copy.cs [hunk 2]
```diff
@@ -229,18 +229,17 @@ private void LogDiagnostic(string message, params object[] messageArgs)
 
             string destinationFolder = Path.GetDirectoryName(destinationFileState.Name);
 
-            if (!string.IsNullOrEmpty(destinationFolder) && !_directoriesKnownToExist.ContainsKey(destinationFolder))
+            if (!string.IsNullOrEmpty(destinationFolder) &&
+                // It's very common for a lot of files to be copied to the same folder. 
+                // Eg., "c:\foo\a"->"c:\bar\a", "c:\foo\b"->"c:\bar\b" and so forth.
+                // We don't want to check whether this folder exists for every single file we copy. So store which we've checked.
+                _directoriesKnownToExist.TryAdd(destinationFolder, true))
```
Comment (Forgind) [inline src/Tasks/Copy.cs:236]: Note the absence of `!`. Finding it in the dictionary means it cannot be added, so ContainsKey and TryAdd return opposite values here.

File: src/Tasks/ManifestUtil/ApplicationManifest.cs [hunk 3]
```diff
@@ -421,25 +421,25 @@ private void ValidateCom()
                             if (!String.IsNullOrEmpty(comInfo.ClsId))
                             {
                                 string key = comInfo.ClsId.ToLowerInvariant();
-                                if (!clsidList.ContainsKey(key))
```
Comment (Forgind) [inline src/Tasks/ManifestUtil/ApplicationManifest.cs:424]: There are numerous cases in which we could eliminate another dictionary access via TryAdd or some similar method, but that isn't available on Framework, sadly.

**Review by Forgind — COMMENTED** [review]
I have a few unrelated changes like in RoslynCodeTaskFactory, but I think it's all fairly straightforward. Also, some of my variable names are bad. If you have better ideas, please suggest them.

File: src/Tasks/Copy.cs (hunk 2 above)
Comment (Forgind) [reply src/Tasks/Copy.cs]: It appears this prevented the creation of some directories where necessary, so I removed it.

**Review by cdmihai — APPROVED** [review]

File: src/Tasks/AssemblyDependency/GlobalAssemblyCache.cs [hunk 4]
```diff
@@ -280,9 +280,8 @@ bool specificVersion
                 }
                 else
                 {
-                    if (fusionNameToResolvedPath.ContainsKey(strongName))
+                    if (fusionNameToResolvedPath.TryGetValue(strongName, out string fusionName))
                     {
-                        fusionNameToResolvedPath.TryGetValue(strongName, out string fusionName);
```
Comment (ladipro) [inline src/Tasks/AssemblyDependency/GlobalAssemblyCache.cs:285]: image

File: src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs (hunk 1 above)
Comment (ladipro) [reply src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs:489]: I would slightly prefer the actual logic to be on its own lines and not part the `VerifyThrow()` call. It makes it easier to read for me when error handling does not mutate the state.

**Review by ladipro — APPROVED** [review]

File: src/Build/Evaluation/Evaluator.cs [hunk 5]
```diff
@@ -849,14 +849,10 @@ private void PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
...
-                                _projectSupportsReturnsAttribute[currentProjectOrImport] |= (target.Returns != null);
-                            }
-                            else
-                            {
-                                _projectSupportsReturnsAttribute[currentProjectOrImport] = (target.Returns != null);
-                            }
+                            // Defaults to false
+                            _projectSupportsReturnsAttribute.TryGetValue(currentProjectOrImport, out NGen<bool> projectSupportsReturnsAttribute);
+
+                            _projectSupportsReturnsAttribute[currentProjectOrImport] = projectSupportsReturnsAttribute | (target.Returns != null);
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/Evaluation/Evaluator.cs:855]: I was just maintaining how it was before, but I agree that's more natural. Thanks!

File: src/Build/Construction/Solution/SolutionProjectGenerator.cs [hunk 6]
```diff
@@ -783,7 +783,7 @@ private void EvaluateAndAddProjects(List<ProjectInSolution> projectsInOrder, Lis
                 AddTraversalTargetForProject(traversalInstance, project, projectConfiguration, "Publish", null, canBuildDirectly);
 
                 // Add any other targets specified by the user that were not already added
-                foreach (string targetName in _targetNames.Where(i => !traversalInstance.Targets.ContainsKey(i)))
+                foreach (string targetName in _targetNames.Except(traversalInstance.Targets.Keys, StringComparer.OrdinalIgnoreCase))
```
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/Construction/Solution/SolutionProjectGenerator.cs:786]: I'd initially had it case-sensitive, and it failed in a way that made this look guilty, but I found a couple other bugs since then. Let me try again and get back to you.

[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/Construction/Solution/SolutionProjectGenerator.cs:786]: SolutionProjectGenerator_Tests.IllegalUserTargetNamesDoNotThrow(forceCaseDifference: True) was the test that failed. Looking at it, it explicitly changes the casing to be different and still expects it to work the same. I would assume that if we should do case-insensitive comparisons at one point, we should do it everywhere in SolutionProjectGenerator when looking at targets' names, though I'm open to being wrong.
